### ISSUE TITLE
Find Variable Features w/ BP Cells matrix

### ISSUE THREAD

**Issue by andrewdchen** [satijalab/seurat#7183]
When I try to run `FindVariableFeatures` on my dataset I get the following error.

```r
object <- FindVariableFeatures(object, verbose = FALSE)
Error: C++ error constructing MatrixTransform: fit.row_params.cols() != loader->rows()
```

I am trying to follow this vignette with my own dataset that has about 9k features by 800k cells. 

My session info:

```r
> sessionInfo()
R version 4.2.1 (2022-06-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /share/pkg.7/r/4.2.1/install/lib64/R/lib/libRblas.so
LAPACK: /share/pkg.7/r/4.2.1/install/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] patchwork_1.1.2.9000    ggrepel_0.9.3           ggplot2_3.4.2           dplyr_1.1.0            
[5] BPCells_0.1.0           Seurat_4.9.9.9041       SeuratObject_4.9.9.9083 sp_1.6-0               

loaded via a namespace (and not attached):
  [1] Rtsne_0.16             colorspace_2.1-0       deldir_1.0-6           ellipsis_0.3.2        
  [5] ggridges_0.5.4         XVector_0.38.0         GenomicRanges_1.50.2   RcppHNSW_0.3.0        
  [9] rstudioapi_0.14        spatstat.data_3.0-1    leiden_0.4.3           listenv_0.9.0         
 [13] RSpectra_0.16-1        fansi_1.0.4            codetools_0.2-18       splines_4.2.1         
 [17] polyclip_1.10-4        spam_2.8-0             jsonlite_1.8.4         ica_1.0-3             
 [21] cluster_2.1.3          png_0.1-8              uwot_0.1.14            shiny_1.7.4           
 [25] sctransform_0.3.5      spatstat.sparse_3.0-1  compiler_4.2.1         httr_1.4.5            
 [29] Matrix_1.5-4           fastmap_1.1.1          lazyeval_0.2.2         cli_3.6.1             
 [33] later_1.3.0            htmltools_0.5.5        tools_4.2.1            igraph_1.4.1          
 [37] dotCall64_1.0-1        GenomeInfoDbData_1.2.9 gtable_0.3.3           glue_1.6.2            
 [41] RANN_2.6.1             reshape2_1.4.4         Rcpp_1.0.10            scattermore_0.8       
 [45] vctrs_0.6.1            spatstat.explore_3.1-0 nlme_3.1-158           progressr_0.13.0      
 [49] lmtest_0.9-40          spatstat.random_3.1-4  stringr_1.5.0          globals_0.16.2        
 [53] mime_0.12              miniUI_0.1.1.1         lifecycle_1.0.3        irlba_2.3.5.1         
 [57] goftest_1.2-3          future_1.32.0          zlibbioc_1.44.0        MASS_7.3-57           
 [61] zoo_1.8-11             scales_1.2.1           promises_1.2.0.1       spatstat.utils_3.0-2  
 [65] parallel_4.2.1         RColorBrewer_1.1-3     reticulate_1.28        pbapply_1.7-0         
 [69] gridExtra_2.3          stringi_1.7.12         S4Vectors_0.36.2       fastDummies_1.6.3     
 [73] BiocGenerics_0.44.0    GenomeInfoDb_1.34.9    bitops_1.0-7           rlang_1.1.0           
 [77] pkgconfig_2.0.3        matrixStats_0.63.0     lattice_0.20-45        ROCR_1.0-11           
 [81] purrr_1.0.1            tensor_1.5             htmlwidgets_1.6.2      cowplot_1.1.1         
 [85] tidyselect_1.2.0       parallelly_1.35.0      RcppAnnoy_0.0.20       plyr_1.8.8            
 [89] magrittr_2.0.3         R6_2.5.1               IRanges_2.32.0         generics_0.1.3        
 [93] withr_2.5.0            pillar_1.9.0           fitdistrplus_1.1-8     RCurl_1.98-1.7        
 [97] survival_3.3-1         abind_1.4-7            tibble_3.2.0           future.apply_1.10.0   
[101] KernSmooth_2.23-20     utf8_1.2.3             spatstat.geom_3.1-0    plotly_4.10.1         
[105] grid_4.2.1             data.table_1.14.8      digest_0.6.31          xtable_1.8-6          
[109] tidyr_1.3.0            httpuv_1.6.9           stats4_4.2.1           munsell_0.5.0         
[113] viridisLite_0.4.1  
```

**Comment by yuhanH** [comment]
Do you meet similar issues if you run FindVariableFeatures in pbmc3k in SeuratData?
```

library(Seurat)
options(Seurat.object.assay.version = 'v5')
library(BPCells)
library(SeuratData)

pbmc3k <- LoadData('pbmc3k')
# convert sparse matrix into BPCells matrix
pbmc3k[['RNA']]$counts <- write_matrix_memory(pbmc3k[['RNA']]$counts)
pbmc3k <- FindVariableFeatures(pbmc3k)

```

**Comment by yuhanH** [comment]
As we have not received a response from you in the past two weeks, I will proceed to close this issue. If the issue is not solved, I will report it.

**Comment by ramonvidal** [comment]
Hi, I have similar issue using [this](https://satijalab.org/seurat/articles/parsebio_sketch_integration) vignete from Seurat and same input dataset.

```
object <- FindVariableFeatures(object)
Finding variable features for layer counts.H_4579
Error: C++ error constructing MatrixTransform: fit.row_params.cols() != loader->rows()
```



```
> sessionInfo()
R version 4.3.1 (2023-06-16 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=English_Germany.utf8  LC_CTYPE=English_Germany.utf8    LC_MONETARY=English_Germany.utf8
[4] LC_NUMERIC=C                     LC_TIME=English_Germany.utf8    

time zone: Europe/Berlin
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] pbmc3k.SeuratData_3.1.4 SeuratData_0.2.2.9001   patchwork_1.1.2         ggrepel_0.9.3          
 [5] ggplot2_3.4.2           dplyr_1.1.2             BPCells_0.1.0           Seurat_4.9.9.9049      
 [9] SeuratObject_4.9.9.9085 sp_2.0-0                Matrix_1.5-4.1         

loaded via a namespace (and not attached):
  [1] deldir_1.0-9           pbapply_1.7-2          gridExtra_2.3          rlang_1.1.1            magrittr_2.0.3        
  [6] RcppAnnoy_0.0.20       matrixStats_1.0.0      ggridges_0.5.4         compiler_4.3.1         spatstat.geom_3.2-1   
 [11] png_0.1-8              vctrs_0.6.3            reshape2_1.4.4         stringr_1.5.0          crayon_1.5.2          
 [16] pkgconfig_2.0.3        fastmap_1.1.1          ellipsis_0.3.2         utf8_1.2.3             promises_1.2.0.1      
 [21] purrr_1.0.1            jsonlite_1.8.5         goftest_1.2-3          later_1.3.1            DelayedArray_0.26.3   
 [26] spatstat.utils_3.0-3   irlba_2.3.5.1          parallel_4.3.1         cluster_2.1.4          R6_2.5.1              
 [31] ica_1.0-3              stringi_1.7.12         RColorBrewer_1.1-3     spatstat.data_3.0-1    reticulate_1.30       
 [36] parallelly_1.36.0      lmtest_0.9-40          scattermore_1.2        Rcpp_1.0.10            tensor_1.5            
 [41] future.apply_1.11.0    zoo_1.8-12             IRanges_2.34.1         sctransform_0.3.5      httpuv_1.6.11         
 [46] splines_4.3.1          igraph_1.5.0           tidyselect_1.2.0       rstudioapi_0.14        abind_1.4-5           
 [51] spatstat.random_3.1-5  codetools_0.2-19       miniUI_0.1.1.1         spatstat.explore_3.2-1 listenv_0.9.0         
 [56] lattice_0.21-8         tibble_3.2.1           plyr_1.8.8             withr_2.5.0            shiny_1.7.4           
 [61] ROCR_1.0-11            Rtsne_0.16             future_1.32.0          fastDummies_1.6.3      survival_3.5-5        
 [66] polyclip_1.10-4        useful_1.2.6           fitdistrplus_1.1-11    pillar_1.9.0           MatrixGenerics_1.12.2 
 [71] KernSmooth_2.23-21     stats4_4.3.1           plotly_4.10.2          generics_0.1.3         RcppHNSW_0.4.1        
 [76] S4Vectors_0.38.1       munsell_0.5.0          scales_1.2.1           globals_0.16.2         xtable_1.8-4          
 [81] glue_1.6.2             lazyeval_0.2.2         tools_4.3.1            data.table_1.14.8      RSpectra_0.16-1       
 [86] RANN_2.6.1             fs_1.6.2               leiden_0.4.3           dotCall64_1.0-2        cowplot_1.1.1         
 [91] grid_4.3.1             tidyr_1.3.0            colorspace_2.1-0       nlme_3.1-162           cli_3.6.1             
 [96] rappdirs_0.3.3         spatstat.sparse_3.0-2  spam_2.9-1             fansi_1.0.4            S4Arrays_1.0.4        
[101] viridisLite_0.4.2      uwot_0.1.15            gtable_0.3.3           digest_0.6.32          BiocGenerics_0.46.0   
[106] progressr_0.13.0       htmlwidgets_1.6.2      htmltools_0.5.5        lifecycle_1.0.3        httr_1.4.6            
[111] mime_0.12              MASS_7.3-60   
```

**Comment by westonelison** [comment]
Hi! I wanted to follow up on this. I am running into the same error when trying to run FIndVariableFeatures on a dataset with four layers and a total of just over a million cells. I ran the code provided by @yuhanH without any trouble.

`> adata <- FindVariableFeatures(adata, verbose = FALSE)

Error in eval(expr, envir, enclos): C++ error constructing MatrixTransform: fit.row_params.cols() != loader->rows()
Traceback:

1. FindVariableFeatures(adata, verbose = FALSE)
2. FindVariableFeatures.Seurat(adata, verbose = FALSE)
3. FindVariableFeatures(object = object[[assay]], selection.method = selection.method, 
 .     loess.span = loess.span, clip.max = clip.max, mean.function = mean.function, 
 .     dispersion.function = dispersion.function, num.bin = num.bin, 
 .     binning.method = binning.method, nfeatures = nfeatures, nselect = nfeatures, 
 .     mean.cutoff = mean.cutoff, dispersion.cutoff = dispersion.cutoff, 
 .     verbose = verbose, ...)
4. FindVariableFeatures.StdAssay(object = object[[assay]], selection.method = selection.method, 
 .     loess.span = loess.span, clip.max = clip.max, mean.function = mean.function, 
 .     dispersion.function = dispersion.function, num.bin = num.bin, 
 .     binning.method = binning.method, nfeatures = nfeatures, nselect = nfeatures, 
 .     mean.cutoff = mean.cutoff, dispersion.cutoff = dispersion.cutoff, 
 .     verbose = verbose, ...)
5. hvf.function(object = data, method = method, nselect = nselect, 
 .     span = span, clip = clip, verbose = verbose, ...)
6. FindVariableFeatures.default(object = data, method = method, 
 .     nselect = nselect, span = span, clip = clip, verbose = verbose, 
 .     ...)
7. method(data = object, nselect = nselect, verbose = verbose, ...)
8. VST.IterableMatrix(data = object, nselect = nselect, verbose = verbose, 
 .     ...)
9. BPCells::matrix_stats(matrix = data.standard, row_stats = "variance")
10. iterate_matrix(matrix)
11. iterate_matrix(matrix)
12. iterate_matrix(x@matrix)
13. iterate_matrix(x@matrix)
14. iterate_matrix_min_by_row_cpp(iterate_matrix(x@matrix), x@row_params)`

`>sessionInfo()

R version 4.2.3 (2023-03-15)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS/LAPACK: /home/welison/.conda/envs/seurat5_bpcells_2/lib/libopenblasp-r0.3.24.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] pbmc3k.SeuratData_3.1.4   SeuratData_0.2.2.9001    
 [3] biomaRt_2.54.1            Azimuth_0.4.6.9004       
 [5] shinyBS_0.61.1            BPCells_0.1.0            
 [7] SeuratWrappers_0.3.19     gridExtra_2.3            
 [9] data.table_1.14.8         harmony_1.0              
[11] Rcpp_1.0.11               Matrix_1.6-1             
[13] ggplot2_3.4.3             dplyr_1.1.3              
[15] EnsDb.Hsapiens.v86_2.99.0 ensembldb_2.22.0         
[17] AnnotationFilter_1.22.0   GenomicFeatures_1.50.4   
[19] AnnotationDbi_1.60.2      Biobase_2.58.0           
[21] GenomicRanges_1.50.2      GenomeInfoDb_1.34.9      
[23] IRanges_2.32.0            S4Vectors_0.36.2         
[25] BiocGenerics_0.44.0       Signac_1.11.9000         
[27] Seurat_4.9.9.9059         SeuratObject_4.9.9.9091  
[29] sp_2.0-0                  hdf5r_1.3.8              
[31] reticulate_1.31          

loaded via a namespace (and not attached):
  [1] rappdirs_0.3.3                    rtracklayer_1.58.0               
  [3] pbdZMQ_0.3-10                     scattermore_1.2                  
  [5] R.methodsS3_1.8.2                 tidyr_1.3.0                      
  [7] JASPAR2020_0.99.10                bit64_4.0.5                      
  [9] irlba_2.3.5.1                     DelayedArray_0.24.0              
 [11] R.utils_2.12.2                    KEGGREST_1.38.0                  
 [13] TFBSTools_1.36.0                  RCurl_1.98-1.12                  
 [15] generics_0.1.3                    cowplot_1.1.1                    
 [17] RSQLite_2.3.1                     RANN_2.6.1                       
 [19] future_1.33.0                     bit_4.0.5                        
 [21] tzdb_0.4.0                        spatstat.data_3.0-1              
 [23] xml2_1.3.5                        httpuv_1.6.11                    
 [25] SummarizedExperiment_1.28.0       DirichletMultinomial_1.40.0      
 [27] gargle_1.5.2                      hms_1.1.3                        
 [29] evaluate_0.21                     promises_1.2.1                   
 [31] fansi_1.0.4                       restfulr_0.0.15                  
 [33] progress_1.2.2                    caTools_1.18.2                   
 [35] dbplyr_2.3.3                      igraph_1.5.1                     
 [37] DBI_1.1.3                         htmlwidgets_1.6.2                
 [39] spatstat.geom_3.2-5               googledrive_2.1.1                
 [41] purrr_1.0.2                       ellipsis_0.3.2                   
 [43] RSpectra_0.16-1                   annotate_1.76.0                  
 [45] deldir_1.0-9                      MatrixGenerics_1.10.0            
 [47] vctrs_0.6.3                       remotes_2.4.2.1                  
 [49] SeuratDisk_0.0.0.9020             ROCR_1.0-11                      
 [51] abind_1.4-5                       cachem_1.0.8                     
 [53] withr_2.5.0                       BSgenome.Hsapiens.UCSC.hg38_1.4.5
 [55] BSgenome_1.66.3                   progressr_0.14.0                 
 [57] presto_1.0.0                      sctransform_0.3.5                
 [59] GenomicAlignments_1.34.1          prettyunits_1.1.1                
 [61] goftest_1.2-3                     cluster_2.1.4                    
 [63] IRdisplay_1.1                     dotCall64_1.0-2                  
 [65] lazyeval_0.2.2                    seqLogo_1.64.0                   
 [67] crayon_1.5.2                      spatstat.explore_3.2-3           
 [69] pkgconfig_2.0.3                   nlme_3.1-163                     
 [71] ProtGenerics_1.30.0               rlang_1.1.1                      
 [73] globals_0.16.2                    lifecycle_1.0.3                  
 [75] miniUI_0.1.1.1                    filelock_1.0.2                   
 [77] fastDummies_1.7.3                 BiocFileCache_2.6.1              
 [79] rsvd_1.0.5                        cellranger_1.1.0                 
 [81] polyclip_1.10-4                   RcppHNSW_0.4.1                   
 [83] matrixStats_1.0.0                 lmtest_0.9-40                    
 [85] IRkernel_1.3.2                    Rhdf5lib_1.20.0                  
 [87] zoo_1.8-12                        base64enc_0.1-3                  
 [89] ggridges_0.5.4                    googlesheets4_1.1.1              
 [91] png_0.1-8                         viridisLite_0.4.2                
 [93] rjson_0.2.21                      shinydashboard_0.7.2             
 [95] bitops_1.0-7                      R.oo_1.25.0                      
 [97] rhdf5filters_1.10.1               KernSmooth_2.23-22               
 [99] spam_2.9-1                        Biostrings_2.66.0                
[101] blob_1.2.4                        stringr_1.5.0                    
[103] parallelly_1.36.0                 spatstat.random_3.1-5            
[105] readr_2.1.4                       CNEr_1.34.0                      
[107] scales_1.2.1                      memoise_2.0.1                    
[109] magrittr_2.0.3                    plyr_1.8.8                       
[111] ica_1.0-3                         zlibbioc_1.44.0                  
[113] compiler_4.2.3                    BiocIO_1.8.0                     
[115] RColorBrewer_1.1-3                fitdistrplus_1.1-11              
[117] Rsamtools_2.14.0                  cli_3.6.1                        
[119] XVector_0.38.0                    listenv_0.9.0                    
[121] patchwork_1.1.3                   pbapply_1.7-2                    
[123] MASS_7.3-60                       tidyselect_1.2.0                 
[125] stringi_1.7.12                    yaml_2.3.7                       
[127] ggrepel_0.9.3                     grid_4.2.3                       
[129] fastmatch_1.1-4                   tools_4.2.3                      
[131] future.apply_1.11.0               parallel_4.2.3                   
[133] uuid_1.1-1                        TFMPvalue_0.0.9                  
[135] Rtsne_0.16                        digest_0.6.33                    
[137] BiocManager_1.30.22               shiny_1.7.5                      
[139] pracma_2.4.2                      later_1.3.1                      
[141] RcppAnnoy_0.0.21                  httr_1.4.7                       
[143] colorspace_2.1-0                  fs_1.6.3                         
[145] XML_3.99-0.14                     tensor_1.5                       
[147] splines_4.2.3                     uwot_0.1.16                      
[149] RcppRoll_0.3.0                    spatstat.utils_3.0-3             
[151] plotly_4.10.2                     xtable_1.8-4                     
[153] jsonlite_1.8.7                    poweRlaw_0.70.6                  
[155] R6_2.5.1                          pillar_1.9.0                     
[157] htmltools_0.5.6                   mime_0.12                        
[159] glue_1.6.2                        fastmap_1.1.1                    
[161] DT_0.29                           BiocParallel_1.32.6              
[163] codetools_0.2-19                  utf8_1.2.3                       
[165] lattice_0.21-8                    spatstat.sparse_3.0-2            
[167] tibble_3.2.1                      curl_5.0.2                       
[169] leiden_0.4.3                      gtools_3.9.4                     
[171] shinyjs_2.1.0                     GO.db_3.16.0                     
[173] survival_3.5-7                    repr_1.1.6                       
[175] munsell_0.5.0                     rhdf5_2.42.1                     
[177] GenomeInfoDbData_1.2.9            reshape2_1.4.4                   
[179] gtable_0.3.4                     `

**Comment by westonelison** [comment]
For those that run into this same issue:

I made it past this error. I missed the hidden section in the interacting with BPCells Vignette that talks about converting an existing Seurat object into one using on-disk storage. Instead I had been trying to save out the matrix back to a hda5 and load in using BPCells. For whatever reason I did not run into this error when I directly converted my Seurat object assay to Assay5 and switched the matrix to on-disk as in the vignette.
